### PR TITLE
chore: support api v4 (development)

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

- v4 is the new development API so we need to add support for it. MLS is only supported on V4 so I'm also bumping the minimum required version for MLS to API v4.

- ~V3 is now stable but I won't mark it as supported in this PR so for production builds this PR should have no changes.~

- V3 is now marked as supported

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/v4/authenticated/ConversationApiV4.kt [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+                emptySet(),
+                response.receiptMode
+            )
+        }
+    }
+
+    override suspend fun updateAccess(
+        conversationId: ConversationId,
+        updateConversationAccessRequest: UpdateConversationAccessRequest
+    ): NetworkResponse<UpdateConversationAccessResponse> = try {
```
Comment (mchenani) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/v4/authenticated/ConversationApiV4.kt:123]: why not wrapping it with `wrapKaliumResponse`?

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/v4/authenticated/NotificationApiV4.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    private val httpClient get() = authenticatedNetworkClient.httpClient
+
+    override suspend fun notificationsCall(
+        querySize: Int,
+        queryClient: String,
+        querySince: String?
+    ): NetworkResponse<NotificationResponse> = wrapKaliumResponse {
+        // Pretty much the same V0 request, but without the 404 overwrite
+        httpClient.get(V0.PATH_NOTIFICATIONS) {
```
Comment (mchenani) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/v4/authenticated/NotificationApiV4.kt:45]: isn't better to have the variables defined for each api version class? here we using the V0 variables.

**Review by mchenani — COMMENTED** [review]
LGTM, but I have some questions and comments.

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/v4/authenticated/NotificationApiV4.kt (hunk 2 above)
Comment (typfel) [reply network/src/commonMain/kotlin/com/wire/kalium/network/api/v4/authenticated/NotificationApiV4.kt]: Yes, but I'm not implementing anything new here. I'm adding api V4 by adding the necessary scaffolding. I'm not touching the implementation.

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/v4/authenticated/MLSMessageApiV4.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    private val authenticatedNetworkClient: AuthenticatedNetworkClient
+) : MLSMessageApiV3(authenticatedNetworkClient) {
+
+    private val httpClient get() = authenticatedNetworkClient.httpClient
+
+    override suspend fun sendCommitBundle(bundle: MLSMessageApi.CommitBundle): NetworkResponse<SendMLSMessageResponse> =
+        wrapKaliumResponse {
+            httpClient.post(PATH_COMMIT_BUNDLES) {
+                setBody(bundle.value)
+                contentType(ContentType.Application.XProtoBuf)
```
[the remark this answers is not in the document]
Comment (typfel) [reply network/src/commonMain/kotlin/com/wire/kalium/network/api/v4/authenticated/MLSMessageApiV4.kt:60]: Maybe, I can try removing it in follow up PR

**Review by vitorhugods — APPROVED** [review]

**Review by MohamadJaara — APPROVED** [review]
LGTM

**Pull request merged**